services/web/src/components/form-fields/Roles.js

```diff
@@ -1,269 +1,302 @@
+          scope,
+          role,
+          scopeRef,
+          roleDefinition: roles[role],
         };
-      })
-    );
-    this.setState({
-      roles,
-    });
-    this.props.onChange(roles);
-  }
+      });
+      this.props.onChange(evt, { value, ...rest });
+    }
+  };
 
-  getGlobalValues() {
-    return this.state.roles
-      .filter((role) => role.scope === 'global')
-      .map((role) => role.role);
+  getValues() {
+    if (this.state.loading) {
+      return [];
+    }
+    return this.props.value.map((role) => this.getRoleId(role));
   }
 
-  getScopedValues(scope, scopeRef) {
-    return this.state.roles
-      .filter((role) => role.scope === scope && role.scopeRef === scopeRef)
-      .map((role) => role.role);
+  getOptions() {
+    return [
+      ...this.getGlobalOptions(),
+      ...this.getOrganizationOptions(),
+      {
+        text: 'Add Organization Role',
+        value: 'add',
+      },
+    ];
   }
 
-  setScopedValues(scope, scopeRef, values) {
-    const otherRoles = this.state.roles.filter(
-      (role) => !(role.scope === scope && role.scopeRef === scopeRef)
-    );
-    const roles = otherRoles.concat(
-      values.map((value) => {
+  getGlobalOptions() {
+    const { roles } = this.state;
+    return Object.entries(roles || {})
+      .filter(([, role]) => {
+        return role.allowScopes.includes('global');
+      })
+      .map(([id, role]) => {
         return {
-          scope,
-          scopeRef,
-          role: value,
+          value: `global:${id}`,
+          text: (
+            <React.Fragment>
+              <Icon name="globe" />
+              {role.name}
+            </React.Fragment>
+          ),
         };
+      });
+  }
+
+  getOrganizationOptions() {
+    if (this.state.loading) {
+      return [];
+    }
+    const { value } = this.props;
+    const { organizations } = this.state;
+    return value
+      .filter((role) => {
+        return role.scope === 'organization';
       })
-    );
-    this.setState({
-      roles,
-    });
-    this.props.onChange(roles);
+      .map((role) => {
+        const org = organizations.find((org) => {
+          return org.id === role.scopeRef;
+        });
+        return {
+          value: this.getRoleId(role),
+          text: (
+            <React.Fragment>
+              <Icon name="building" />
+              {org.name} {role.roleDefinition.name}
+            </React.Fragment>
+          ),
+        };
+      });
   }
 
-  getScopeRefs(roles, scope) {
-    return union(
-      roles.filter((role) => role.scope === scope).map((role) => role.scopeRef)
-    );
+  getRoleId(role) {
+    return [role.scope, role.role, role.scopeRef]
+      .filter((str) => str)
+      .join(':');
   }
 
-  fetchOrganizations = async () => {
-    const { data } = await request({
-      method: 'POST',
-      path: '/1/organizations/search',
-      body: {},
+  // Organization search dropdown
+
+  onSearchChange = (evt, { value: organization }) => {
```

Comment:
just wondering why you didn't do this inline e.g. its kind of bound to the search dropdown.

Reply:
is that preferred? I more prefer better names and less inlining... there's also the geek argument that inlining creates a new function on each render 🤓

Reply:
My thinking on this (totally opinionated and really perhaps more subjective aesthetics and style - mixed bag of 💩 )

Keep things that can easy be read by the user (~2-3 lines) together.
Example. `onClick={someaction}` if someaction is small enough to be inline, then it's easy for the user understand the code, as they don't have to scroll around to find what `someaction` does.

Keep functions on the classes, that are reuse-able  (unless you are trying to optimize or its overly complex for some reason) 

Try to keep the functions argument list free of `dom` event, to make sure it's not weird to call from a regular function.
e.g. calling this. `onSearchChange({}, { value: "bob" })` is somewhat weird while `setOrganization("bob")` is somewhat more clear. 

Fine to repeat yourself (in small scale)

The greek argument, yeah I wouldn't worry about it, unless speed is what we are looking for. Also tell me is that only valid if the child component is a pure component (or a function component using hooks)?

Reply:
left this for now but noted for the future